agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -197,52 +190,99 @@ public Function thenSort(@NonNull final Comparator comparator) {
+        final Object command = functions[i++];
+        switch (type) {
+          case LIMIT: {
+            final List list = (List) item;
+            final int limit = (int) command;
+            if (limit <= 0) {
+              item = emptyList();
+            } else if (list.size() < limit) {
+              item = input;
+            } else if (compilerList) {
+              while (list.size() > limit) {
+                list.remove(list.size() - 1);
+              }
+            } else {
+              final ArrayList newList = new ArrayList(limit);
+              for (int copied = 0; copied < limit; copied++) {
+                newList.add(list.get(copied));
+              }
+              item = newList;
+              compilerList = true;
+            }
+          }
+          break;
+          case FILTER: {
+            final List list = (List) item;
+            if (list.isEmpty()) {
+              item = emptyList();
+            } else if (compilerList) {
+              final Predicate predicate = (Predicate) command;
+              final int size = list.size();
+              for (int j = size - 1; j >= 0; j--) {
+                final Object listItem = list.get(j);
+                if (!predicate.apply(listItem)) {
+                  list.remove(j);
+                }
+              }
+            } else {
+              final Predicate predicate = (Predicate) command;
+              final int size = list.size();
+              final List result = new ArrayList(size);
+              for (int j = 0; j < size; j++) {
+                final Object listItem = list.get(j);
+                if (predicate.apply(listItem)) {
+                  result.add(listItem);
+                }
+              }
+              item = result;
+              compilerList = true;
+            }
+          }
+          break;
+          case SORT: {
+            final List output = compilerList ? ((List) item) : new ArrayList<>(((List) item));
+            Collections.sort(output, (Comparator) command);
+            item = output;
+            compilerList = true;
+          }
+          break;
+          case MAP: {
+            final List list = (List) item;
+            final int size = list.size();
+            if (list.isEmpty()) {
+              item = emptyList();
+            } else if (compilerList) {
+              final Function function = (Function) command;
+              for (int j = 0; j < size; j++) {
+                list.set(j, function.apply(list.get(j)));
+              }
+            } else {
+              final Function function = (Function) command;
+              final List result = new ArrayList(size);
+              for (int j = 0; j < size; j++) {
+                result.add(function.apply(list.get(j)));
+              }
+              item = result;
+              compilerList = true;
+            }
+          }
+          break;
+          default: {
+            item = ((Function) command).apply(item);
```

Comment:
My reason is only one case apply to item without casting it to List, it could be reasonable to separate it from the reset.

At least make it `case APPLY:` not just `default:`